section/Lua/LuaObject.cxx

```diff
@@ -0,0 +1,285 @@
+    result.AssignNewTable(m_state, t->sizearray, 1 << t->lsizenode);
+    backref.SetObject(*this, result);
+
+    for (const auto &[key, value] : Pairs(*this))
+    {
+        LuaObject self_ref = backref.GetObject(value);
+
+        if (self_ref.IsNil())
+        {
+            LuaObject clonedValue = value.IsTable() ? value.__Clone(backref) : value;
+            result.SetObject(key, clonedValue);
+        }
+        else
+        {
+            result.SetObject(key, self_ref);
+        }
+    }
+    return result;
+}
+
+inline int LuaObject::Type() const
+{
+    luaplus_assert(m_state);
+    return m_object.tt;
+}
+
+inline bool LuaObject::IsBoolean() const { return Type() == LUA_TBOOLEAN; }
+inline bool LuaObject::IsInteger() const { return Type() == LUA_TNUMBER; }
+inline bool LuaObject::IsNumber() const { return Type() == LUA_TNUMBER; }
+inline bool LuaObject::IsString() const { return Type() == LUA_TSTRING; }
+inline bool LuaObject::IsTable() const { return Type() == LUA_TTABLE; }
+inline bool LuaObject::IsNil() const { return m_state && Type() == LUA_TNIL; }
+inline bool LuaObject::IsUserData() const
+{
+    return Type() == LUA_TUSERDATA || Type() == LUA_TLIGHTUSERDATA;
+}
+
+void LuaObject::RemoveFromUsedList()
+{
+    if (m_state)
+    {
+        m_prev->m_next = m_next;
+        m_next->m_prev = m_prev;
+        m_object.tt = LUA_TNIL;
+    }
+}
+
+void LuaObject::Reset()
+{
+    RemoveFromUsedList();
+    m_state = nullptr;
+}
+
+const char *LuaObject::TypeName() const
+{
+    int tt = Type();
+    if (tt == LUA_TNONE)
+    {
+        return "no value";
+    }
+    return luaT_typenames[tt];
+}
+
+void LuaObject::SetTableHelper(const TObject *key, const TObject *value) const
+{
+    luaV_settable(GetActiveCState(), &m_object, key, value);
+}
+
+void LuaObject::SetNil(const LuaObject &key) const
+{
+    luaplus_assert(m_state == key.m_state);
+    TObject value{};
+    SetTableHelper(&key.m_object, &value);
+}
+
+void LuaObject::SetObject(const LuaObject &key, const LuaObject &value) const
+{
+    luaplus_assert(m_state == key.m_state);
+    luaplus_assert(m_state == value.m_state);
+    SetTableHelper(&key.m_object, &value.m_object);
+}
+
+// void LuaObject::SetObject(const char *key, const LuaObject &value) const {
+//   TObject key_obj{};
+//   key_obj.tt = LUA_TSTRING;
+// }
+
+void LuaObject::SetObject(int key, const LuaObject &value) const
+{
+    TObject key_obj{key};
+    SetTableHelper(&key_obj, &value.m_object);
+}
+
+inline const TObject *LuaObject::GetTableHelper(const TObject *key) const
+{
+    return luaV_gettable(GetActiveCState(), &m_object, key, 0);
+}
+
+LuaObject LuaObject::GetObject(const LuaObject &key) const
+{
+    luaplus_assert(m_state == key.m_state);
+    return LuaObject{m_state, GetTableHelper(&key.m_object)};
+}
+
+LuaObject LuaObject::GetObject(int key) const
+{
+    TObject key_obj{key};
+    return LuaObject{m_state, GetTableHelper(&key_obj)};
+}
+
+lua_Number LuaObject::GetNumber(int key) const
```

Comment:
I have these methods as `GetByObject` and `GetByIndex` - these method names don't take an argument

Reply:
This is a new one to retrieve number by index.

Reply:
Oh, I thought you misnamed the versions LuaPlus provided. That's fine I guess.